// @microsoft/teamsfx-core/build/plugins/solution/fx-solution/question.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserEmailQuestion = exports.AskSubscriptionQuestion = exports.DeployPluginSelectQuestion = exports.addCapabilityQuestion = exports.createAddAzureResourceQuestion = exports.AzureResourcesQuestion = exports.BotNotificationTriggers = exports.BotScenario = exports.AzureResourceKeyVault = exports.AzureResourceApim = exports.AzureResourceFunction = exports.AzureResourceSQL = exports.HostTypeOptionSPFx = exports.HostTypeOptionAzure = exports.AzureSolutionQuestionNames = exports.M365SearchAppOptionItem = exports.M365SsoLaunchPageOptionItem = exports.TabNonSsoItem = exports.BotSsoItem = exports.TabSsoItem = exports.TabSPFxNewUIItem = exports.TabSPFxItem = exports.MessageExtensionNewUIItem = exports.MessageExtensionItem = exports.ExistingTabOptionItem = exports.CommandAndResponseOptionItem = exports.NotificationOptionItem = exports.BotOptionItem = exports.TabNewUIOptionItem = exports.TabOptionItem = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const featureFlags_1 = require("../../../common/featureFlags");
const localizeUtils_1 = require("../../../common/localizeUtils");
exports.TabOptionItem = {
    id: "Tab",
    label: localizeUtils_1.getLocalizedString("core.TabOption.label"),
    cliName: "tab",
    description: localizeUtils_1.getLocalizedString("core.TabOption.description"),
    detail: localizeUtils_1.getLocalizedString("core.TabOption.detail"),
};
exports.TabNewUIOptionItem = {
    id: "Tab",
    label: localizeUtils_1.getLocalizedString("core.TabOption.labelNew"),
    cliName: "tab",
    detail: localizeUtils_1.getLocalizedString("core.TabOption.detailNew"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.BotOptionItem = {
    id: "Bot",
    label: "Bot",
    cliName: "bot",
    description: localizeUtils_1.getLocalizedString("core.BotOption.description"),
    detail: localizeUtils_1.getLocalizedString("core.BotOption.detail"),
};
exports.NotificationOptionItem = {
    id: "Notification",
    label: localizeUtils_1.getLocalizedString("core.NotificationOption.label"),
    cliName: "notification",
    detail: localizeUtils_1.getLocalizedString("core.NotificationOption.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.CommandAndResponseOptionItem = {
    // For default option, id and cliName must be the same
    id: "command-bot",
    label: localizeUtils_1.getLocalizedString("core.CommandAndResponseOption.label"),
    cliName: "command-bot",
    detail: localizeUtils_1.getLocalizedString("core.CommandAndResponseOption.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.ExistingTabOptionItem = {
    id: "ExistingTab",
    label: localizeUtils_1.getLocalizedString("core.ExistingTabOption.label"),
    cliName: "existing-tab",
    detail: localizeUtils_1.getLocalizedString("core.ExistingTabOption.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.MessageExtensionItem = {
    id: "MessagingExtension",
    label: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.label"),
    cliName: "messaging-extension",
    description: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.description"),
    detail: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.detail"),
};
exports.MessageExtensionNewUIItem = {
    id: "MessagingExtension",
    label: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.labelNew"),
    cliName: "messaging-extension",
    detail: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.TabSPFxItem = {
    id: "TabSPFx",
    label: localizeUtils_1.getLocalizedString("core.TabSPFxOption.label"),
    cliName: "tab-spfx",
    description: localizeUtils_1.getLocalizedString("core.TabSPFxOption.description"),
    detail: localizeUtils_1.getLocalizedString("core.TabSPFxOption.detail"),
};
exports.TabSPFxNewUIItem = {
    id: "TabSPFx",
    label: localizeUtils_1.getLocalizedString("core.TabSPFxOption.labelNew"),
    cliName: "tab-spfx",
    detail: localizeUtils_1.getLocalizedString("core.TabSPFxOption.detailNew"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.TabSsoItem = {
    id: "TabSSO",
    label: "TabSSO",
    cliName: "tab-sso",
    description: localizeUtils_1.getLocalizedString("core.TabSso.description"),
    detail: localizeUtils_1.getLocalizedString("core.TabSso.detail"),
};
exports.BotSsoItem = {
    id: "BotSSO",
    label: "BotSSO",
    cliName: "bot-sso",
    description: localizeUtils_1.getLocalizedString("core.BotSso.description"),
    detail: localizeUtils_1.getLocalizedString("core.BotSso.detail"),
};
exports.TabNonSsoItem = {
    id: "TabNonSso",
    label: localizeUtils_1.getLocalizedString("core.TabNonSso.label"),
    cliName: "tab-non-sso",
    description: localizeUtils_1.getLocalizedString("core.TabNonSso.description"),
    detail: localizeUtils_1.getLocalizedString("core.TabNonSso.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsAppSeparatorOptionItem.label"),
};
exports.M365SsoLaunchPageOptionItem = {
    id: "M365SsoLaunchPage",
    label: "SSO enabled launch page",
    cliName: "sso-launch-page",
    detail: localizeUtils_1.getLocalizedString("core.M365SsoLaunchPageOptionItem.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsM365AppSeparatorOptionItem.label"),
};
exports.M365SearchAppOptionItem = {
    id: "M365SearchApp",
    label: "Search app",
    cliName: "search-app",
    detail: localizeUtils_1.getLocalizedString("core.M365SearchAppOptionItem.detail"),
    groupName: localizeUtils_1.getLocalizedString("core.TeamsM365AppSeparatorOptionItem.label"),
};
var AzureSolutionQuestionNames;
(function (AzureSolutionQuestionNames) {
    AzureSolutionQuestionNames["Capabilities"] = "capabilities";
    AzureSolutionQuestionNames["TabScopes"] = "tab-scopes";
    AzureSolutionQuestionNames["HostType"] = "host-type";
    AzureSolutionQuestionNames["AzureResources"] = "azure-resources";
    AzureSolutionQuestionNames["PluginSelectionDeploy"] = "deploy-plugin";
    AzureSolutionQuestionNames["AddResources"] = "add-azure-resources";
    AzureSolutionQuestionNames["AppName"] = "app-name";
    AzureSolutionQuestionNames["AskSub"] = "subscription";
    AzureSolutionQuestionNames["ProgrammingLanguage"] = "programming-language";
    AzureSolutionQuestionNames["Solution"] = "solution";
    AzureSolutionQuestionNames["Scenarios"] = "scenarios";
})(AzureSolutionQuestionNames = exports.AzureSolutionQuestionNames || (exports.AzureSolutionQuestionNames = {}));
exports.HostTypeOptionAzure = {
    id: "Azure",
    label: localizeUtils_1.getLocalizedString("core.HostTypeOptionAzure.label"),
    cliName: "azure",
};
exports.HostTypeOptionSPFx = {
    id: "SPFx",
    label: localizeUtils_1.getLocalizedString("core.HostTypeOptionSPFx.label"),
    cliName: "spfx",
};
exports.AzureResourceSQL = {
    id: "sql",
    label: localizeUtils_1.getLocalizedString("core.AzureResourceSQL.label"),
    description: localizeUtils_1.getLocalizedString("core.AzureResourceSQL.description"),
};
exports.AzureResourceFunction = {
    id: "function",
    label: localizeUtils_1.getLocalizedString("core.AzureResourceFunction.label"),
};
exports.AzureResourceApim = {
    id: "apim",
    label: localizeUtils_1.getLocalizedString("core.AzureResourceApim.label"),
    description: localizeUtils_1.getLocalizedString("core.AzureResourceApim.description"),
};
exports.AzureResourceKeyVault = {
    id: "keyvault",
    label: localizeUtils_1.getLocalizedString("core.AzureResourceKeyVault.label"),
    description: localizeUtils_1.getLocalizedString("core.AzureResourceKeyVault.description"),
};
var BotScenario;
(function (BotScenario) {
    BotScenario["NotificationBot"] = "notificationBot";
    BotScenario["CommandAndResponseBot"] = "commandAndResponseBot";
})(BotScenario = exports.BotScenario || (exports.BotScenario = {}));
exports.BotNotificationTriggers = {
    Timer: "timer",
    Http: "http",
};
exports.AzureResourcesQuestion = {
    name: AzureSolutionQuestionNames.AzureResources,
    title: "Cloud resources",
    type: "multiSelect",
    staticOptions: [exports.AzureResourceSQL, exports.AzureResourceFunction],
    default: [],
    onDidChangeSelection: async function (currentSelectedIds, previousSelectedIds) {
        if (currentSelectedIds.has(exports.AzureResourceSQL.id)) {
            currentSelectedIds.add(exports.AzureResourceFunction.id);
        }
        return currentSelectedIds;
    },
    placeholder: "Select a resource (optional)",
};
function createAddAzureResourceQuestion(alreadyHaveFunction, alreadyHaveSQL, alreadyHaveAPIM, alreadyHaveKeyVault) {
    const options = [exports.AzureResourceFunction, exports.AzureResourceSQL];
    if (!alreadyHaveAPIM)
        options.push(exports.AzureResourceApim);
    if (!alreadyHaveKeyVault)
        options.push(exports.AzureResourceKeyVault);
    return {
        name: AzureSolutionQuestionNames.AddResources,
        title: "Cloud resources",
        type: "multiSelect",
        staticOptions: options,
        default: [],
        onDidChangeSelection: async function (currentSelectedIds, previousSelectedIds) {
            const hasSQL = currentSelectedIds.has(exports.AzureResourceSQL.id);
            const hasAPIM = currentSelectedIds.has(exports.AzureResourceApim.id);
            if ((hasSQL || hasAPIM) && !alreadyHaveFunction) {
                currentSelectedIds.add(exports.AzureResourceFunction.id);
            }
            return currentSelectedIds;
        },
    };
}
exports.createAddAzureResourceQuestion = createAddAzureResourceQuestion;
function addCapabilityQuestion(alreadyHaveTab, alreadyHaveBot) {
    const options = [];
    if (!alreadyHaveTab)
        options.push(exports.TabOptionItem);
    if (!alreadyHaveBot) {
        options.push(exports.BotOptionItem);
        options.push(exports.MessageExtensionItem);
        options.push(exports.NotificationOptionItem);
        options.push(exports.CommandAndResponseOptionItem);
    }
    return {
        name: AzureSolutionQuestionNames.Capabilities,
        title: featureFlags_1.isBotNotificationEnabled()
            ? localizeUtils_1.getLocalizedString("core.addCapabilityQuestion.titleNew")
            : localizeUtils_1.getLocalizedString("core.addCapabilityQuestion.title"),
        type: "multiSelect",
        staticOptions: options,
        default: [],
    };
}
exports.addCapabilityQuestion = addCapabilityQuestion;
exports.DeployPluginSelectQuestion = {
    name: AzureSolutionQuestionNames.PluginSelectionDeploy,
    title: `Select resources`,
    type: "multiSelect",
    skipSingleOption: true,
    staticOptions: [],
    default: [],
};
exports.AskSubscriptionQuestion = {
    name: AzureSolutionQuestionNames.AskSub,
    type: "func",
    func: async (inputs) => {
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    },
};
function getUserEmailQuestion(currentUserEmail) {
    let defaultUserEmail = "";
    if (currentUserEmail && currentUserEmail.indexOf("@") > 0) {
        defaultUserEmail = "[UserName]@" + currentUserEmail.split("@")[1];
    }
    return {
        name: "email",
        type: "text",
        title: localizeUtils_1.getLocalizedString("core.getUserEmailQuestion.title"),
        default: defaultUserEmail,
        validation: {
            validFunc: (input, previousInputs) => {
                if (!input || input.trim() === "") {
                    return localizeUtils_1.getLocalizedString("core.getUserEmailQuestion.validation1");
                }
                input = input.trim();
                if (input === defaultUserEmail) {
                    return localizeUtils_1.getLocalizedString("core.getUserEmailQuestion.validation2");
                }
                const re = /\S+@\S+\.\S+/;
                if (!re.test(input)) {
                    return localizeUtils_1.getLocalizedString("core.getUserEmailQuestion.validation3");
                }
                return undefined;
            },
        },
    };
}
exports.getUserEmailQuestion = getUserEmailQuestion;
