// @microsoft/teamsfx-core/build/plugins/resource/bot/azureOps.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureOperations = void 0;
const tslib_1 = require("tslib");
const errors_1 = require("./errors");
const strings_1 = require("./resources/strings");
const utils = tslib_1.__importStar(require("./utils/common"));
const axios_1 = tslib_1.__importDefault(require("axios"));
const common_1 = require("../../../common");
const constants_1 = require("./constants");
class AzureOperations {
    static async UpdateBotChannelRegistration(botClient, resourceGroup, botChannelRegistrationName, msaAppId, endpoint, displayName) {
        let botResponse = undefined;
        try {
            botResponse = await botClient.bots.update(resourceGroup, botChannelRegistrationName, {
                properties: {
                    displayName: displayName !== null && displayName !== void 0 ? displayName : botChannelRegistrationName,
                    endpoint: endpoint,
                    msaAppId: msaAppId,
                },
            });
        }
        catch (e) {
            throw new errors_1.MessageEndpointUpdatingError(endpoint, e);
        }
        if (!botResponse || !utils.isHttpCodeOkOrCreated(botResponse._response.status)) {
            throw new errors_1.MessageEndpointUpdatingError(endpoint);
        }
    }
    static async LinkTeamsChannel(botClient, resourceGroup, botChannelRegistrationName) {
        let channelResponse = undefined;
        try {
            channelResponse = await botClient.channels.create(resourceGroup, botChannelRegistrationName, "MsTeamsChannel", {
                location: "global",
                kind: "bot",
                properties: {
                    channelName: "MsTeamsChannel",
                    properties: {
                        isEnabled: true,
                    },
                },
            });
        }
        catch (e) {
            throw new errors_1.ProvisionError(strings_1.CommonStrings.MS_TEAMS_CHANNEL, e);
        }
        if (!channelResponse || !utils.isHttpCodeOkOrCreated(channelResponse._response.status)) {
            throw new errors_1.ProvisionError(strings_1.CommonStrings.MS_TEAMS_CHANNEL);
        }
    }
    static async CreateOrUpdateAzureWebApp(webSiteMgmtClient, resourceGroup, siteName, siteEnvelope, update) {
        let webappResponse = undefined;
        try {
            webappResponse = await webSiteMgmtClient.webApps.createOrUpdate(resourceGroup, siteName, siteEnvelope);
        }
        catch (e) {
            if (!update) {
                throw new errors_1.ProvisionError(strings_1.CommonStrings.AZURE_WEB_APP, e);
            }
            else {
                throw new errors_1.ConfigUpdatingError(strings_1.ConfigNames.AZURE_WEB_APP_AUTH_CONFIGS, e);
            }
        }
        if (!webappResponse || !utils.isHttpCodeOkOrCreated(webappResponse._response.status)) {
            if (!update) {
                throw new errors_1.ProvisionError(strings_1.CommonStrings.AZURE_WEB_APP);
            }
            else {
                throw new errors_1.ConfigUpdatingError(strings_1.ConfigNames.AZURE_WEB_APP_AUTH_CONFIGS);
            }
        }
        return webappResponse;
    }
    static async ListPublishingCredentials(webSiteMgmtClient, resourceGroup, siteName) {
        let listResponse = undefined;
        try {
            listResponse = await webSiteMgmtClient.webApps.listPublishingCredentials(resourceGroup, siteName);
        }
        catch (e) {
            throw new errors_1.ListPublishingCredentialsError(e);
        }
        if (!listResponse || !utils.isHttpCodeOkOrCreated(listResponse._response.status)) {
            throw new errors_1.ListPublishingCredentialsError();
        }
        return listResponse;
    }
    static async ZipDeployPackage(zipDeployEndpoint, zipBuffer, config) {
        let res = undefined;
        try {
            res = await axios_1.default.post(zipDeployEndpoint, zipBuffer, config);
        }
        catch (e) {
            throw new errors_1.ZipDeployError(e);
        }
        if (!res || !utils.isHttpCodeAccepted(res === null || res === void 0 ? void 0 : res.status)) {
            throw new errors_1.ZipDeployError();
        }
        return res.headers.location;
    }
    static async CheckDeployStatus(location, config) {
        let res = undefined;
        for (let i = 0; i < constants_1.DeployStatus.RETRY_TIMES; ++i) {
            try {
                res = await axios_1.default.get(location, config);
            }
            catch (e) {
                throw new errors_1.DeployStatusError(e);
            }
            if (res) {
                if (utils.isHttpCodeAccepted(res === null || res === void 0 ? void 0 : res.status)) {
                    await common_1.waitSeconds(constants_1.DeployStatus.BACKOFF_TIME_S);
                }
                else if (utils.isHttpCodeOkOrCreated(res === null || res === void 0 ? void 0 : res.status)) {
                    return;
                }
                else {
                    throw new errors_1.DeployStatusError();
                }
            }
        }
        throw new errors_1.DeployTimeoutError();
    }
    static async RestartWebApp(webSiteMgmtClient, resourceGroup, siteName) {
        let res = undefined;
        try {
            res = await webSiteMgmtClient.webApps.restart(resourceGroup, siteName);
        }
        catch (e) {
            throw new errors_1.RestartWebAppError(e);
        }
        if (!res || !utils.isHttpCodeOkOrCreated(res === null || res === void 0 ? void 0 : res._response.status)) {
            throw new errors_1.RestartWebAppError();
        }
    }
}
exports.AzureOperations = AzureOperations;
